zerocopy-derive/src/repr.rs

```diff
@@ -6,202 +6,276 @@
 
-use core::fmt::{self, Display, Formatter};
-
-use proc_macro2::Ident;
-use quote::{ToTokens, TokenStreamExt as _};
+use core::{
+    convert::{Infallible, TryFrom},
+    num::NonZeroU64,
+};
 
-use {
-    proc_macro2::Span,
-    syn::punctuated::Punctuated,
-    syn::spanned::Spanned,
-    syn::token::Comma,
-    syn::{Attribute, DeriveInput, Error, LitInt, Meta},
+use proc_macro2::{Span, TokenStream};
+use quote::{quote_spanned, ToTokens, TokenStreamExt as _};
+use syn::{
+    punctuated::Punctuated, spanned::Spanned as _, token::Comma, Attribute, Error, LitInt, Meta,
+    MetaList,
 };
 
-pub(crate) struct Config<Repr: KindRepr> {
-    // A human-readable message describing what combinations of representations
-    // are allowed. This will be printed to the user if they use an invalid
-    // combination.
-    pub(crate) allowed_combinations_message: &'static str,
-    // Whether we're checking as part of `derive(Unaligned)`. If not, we can
-    // ignore `repr(align)`, which makes the code (and the list of valid repr
-    // combinations we have to enumerate) somewhat simpler. If we're checking
-    // for `Unaligned`, then in addition to checking against illegal
-    // combinations, we also check to see if there exists a `repr(align(N > 1))`
-    // attribute.
-    pub(crate) derive_unaligned: bool,
-    // Combinations which are valid for the trait.
-    pub(crate) allowed_combinations: &'static [&'static [Repr]],
-    // Combinations which are not valid for the trait, but are legal according
-    // to Rust. Any combination not in this or `allowed_combinations` is either
-    // illegal according to Rust or the behavior is unspecified. If the behavior
-    // is unspecified, it might become specified in the future, and that
-    // specification might not play nicely with our requirements. Thus, we
-    // reject combinations with unspecified behavior in addition to illegal
-    // combinations.
-    pub(crate) disallowed_but_legal_combinations: &'static [&'static [Repr]],
-}
-
-impl<R: KindRepr> Config<R> {
-    /// Validate that `input`'s representation attributes conform to the
-    /// requirements specified by this `Config`.
+/// The computed representation of a type.
+///
+/// This is the result of processing all `#[repr(...)]` attributes on a type, if
+/// any. A `Repr` is only capable of representing legal combinations of
+/// `#[repr(...)]` attributes.
+#[cfg_attr(test, derive(Copy, Clone, Debug, Eq, PartialEq))]
+pub(crate) enum Repr<Prim, Packed, S> {
+    /// `#[repr(transparent)]`
+    Transparent(S),
+    /// A compound representation: `repr(C)`, `repr(Rust)`, or `repr(Int)`
+    /// optionally combined with `repr(packed(...))` or `repr(align(...))`
+    Compound(Spanned<CompoundRepr<Prim>, S>, Option<Spanned<AlignRepr<Packed>, S>>),
+}
+
+/// A compound representation: `repr(C)`, `repr(Rust)`, or `repr(Int)`.
+#[cfg_attr(test, derive(Copy, Clone, Debug, Eq, PartialEq))]
+pub(crate) enum CompoundRepr<Prim> {
+    C,
```

Comment:
Data-carrying enums can be `repr(C, int)`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=555030b1832a3da32661af452c7d5809

That repr combination can only be used on data-carrying enums, otherwise the compiler emits an error. When `C` and `int` reprs are used together, the enum follows C-style enum layout (outer tag) and explicitly sets the tag primitive.

Also, are we missing a test with this repr combination?

Reply:
Yeah, my worry is that the logic around those is super subtle. My intention (which I should have written down) was to just not support them for now and then follow up later to add support if we need to. I'm not aware of any of our users using them.

> Also, are we missing a test with this repr combination?

No, we currently test that they're rejected:

```rust
test!(@error #[repr(C, u8)] => EnumRepr => FromRawReprs(Conflict).into());
```